frontend/src/scenes/trends/trendsLogic.test.ts

```diff
@@ -0,0 +1,164 @@
+import { mockAPI } from 'lib/api.mock'
+import { expectLogic, initKeaTestLogic } from '~/test/kea-test-utils'
+import { trendsLogic } from 'scenes/trends/trendsLogic'
+import { trendsLogicType } from 'scenes/trends/trendsLogicType'
+import { PropertyOperator, TrendResult } from '~/types'
+
+jest.mock('lib/api')
+
+describe('trendsLogic', () => {
+    let logic: BuiltLogic<trendsLogicType>
+
+    mockAPI(async ({ pathname, searchParams }) => {
+        if (pathname === '_preflight/') {
+            return { is_clickhouse_enabled: true }
+        } else if (pathname === 'api/users/@me/') {
+            return { organization: {}, team: { ingested_event: true, completed_snippet_onboarding: true } }
+        } else if (
+            [
+                'api/action/',
+                'api/projects/@current/event_definitions/',
+                'api/users/@me/',
+                'api/dashboard',
+                'api/insight',
+            ].includes(pathname)
+        ) {
+            return { results: [] }
+        } else if (['api/insight/session/', 'api/insight/trend/'].includes(pathname)) {
+            return { result: ['result from api'] }
+        } else {
+            throw new Error(`Unmocked fetch to: ${pathname} with params: ${JSON.stringify(searchParams)}`)
+        }
+    })
+
+    describe('core assumptions', () => {
+        initKeaTestLogic({
+            logic: trendsLogic,
+            props: {},
+            onLogic: (l) => (logic = l),
+        })
+
+        it('loads results on mount', async () => {
+            await expectLogic(logic).toDispatchActions(['loadResults'])
+        })
+    })
+
+    describe('reducers', () => {
+        initKeaTestLogic({
+            logic: trendsLogic,
+            props: {},
+            onLogic: (l) => (logic = l),
+        })
+
+        it('visibilityMap', async () => {
+            const r = {} as TrendResult
+
+            expectLogic(logic, () => {
+                logic.actions.setVisibilityById({ '1': true, '2': false })
+                logic.actions.setVisibilityById({ '8': true, '2': true })
+            }).toMatchValues({ visibilityMap: { 1: true, 2: true, 8: true } })
+
+            expectLogic(logic, () => {
+                logic.actions.setCachedResultsSuccess({ result: [r, r, r, r, r], filters: {} })
+            }).toMatchValues({ visibilityMap: { 0: true, 1: true, 2: true, 3: true, 4: true } })
+
+            expectLogic(logic, () => {
+                logic.actions.loadResultsSuccess({ result: [r, r], filters: {} })
+            }).toMatchValues({ visibilityMap: { 0: true, 1: true } })
+
+            expectLogic(logic, () => {
+                logic.actions.toggleVisibility(1)
+            }).toMatchValues({ visibilityMap: { 0: true, 1: false } })
+
+            expectLogic(logic, () => {
+                logic.actions.toggleVisibility(1)
+            }).toMatchValues({ visibilityMap: { 0: true, 1: true } })
+        })
+    })
+
+    describe('as dashboard item', () => {
```

Comment:
Does this dashboard test fit better in a insightLogic.test or dashboardItemLogic.test? I think there's a lot of duplicate refresh test code here that's applicable to all insights, and might make more sense given that this is testing dashboard item specific behavior.

Reply:
Absolutely! There's a lot of duplicate code here, and even that is actually not enough. This should also be replicated into `pathsLogic` and `retentionTableLogic`, the other two insights that we have. I decided to not scope creep this PR even further and fix this later.

Just like we have the same test code for both clickhouse and postgres, we should ideally have the same test code for all different insight logics. Taking this a step further, we could even abstract away the shared code into its own logic, and have `trendsLogic` and friends use kea's [undocumented `inherit` feature](https://github.com/keajs/kea/blob/master/CHANGELOG.md#242---2021-05-17) to automatically inherit all actions/methods from it, such as the `results`/`setCachedResults`/etc handling.

I planned to mention this as part of a larger tech debt issue "fix spaghetti logic" I'll make next week (I'm off today actually).